CMakeLists.txt

```diff
@@ -0,0 +1,272 @@
+        if (NETWORKIT_BUILD_TESTS)
+            add_custom_target(networkit_tests_${modname})
+        endif()
+    endif()
+
+
+    # Add source files (it's important to mark them private; otherwise
+    # all targets linking to the lib, will recompile the objects from scratch)
+    foreach(file ${ARGN})
+        target_sources(${MODULE_TARGET}
+            PRIVATE ${CMAKE_CURRENT_LIST_DIR}/${file})
+    endforeach()
+endfunction()
+
+# Analogous to target_link_libraries with KEYWORDS.
+# Use the module's name (without networkit_ prefix) for target.
+# In case of monolithic builds, the call is ignored.
+# To link against another module use networkit_module_link_modules
+# Example: networkit_module_link_libraries(graph PRIVATE foobar_lib)
+function(networkit_module_link_libraries modname)
+    set(options )
+    set(oneValueArgs )
+    set(multiValueArgs PRIVATE PUBLIC)
+    cmake_parse_arguments(NMLL
+            "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    if (NOT NETWORKIT_MONOLITH)
+        target_link_libraries(networkit_${modname}
+                PRIVATE ${NMLL_PRIVATE}
+                PUBLIC ${NMLL_PUBLIC})
+    endif()
+endfunction()
+
+# Specifiy inter-module dependencies. The function expects a list of at least
+# two module names (without the networkit_ prefix). The first one add all following
+# ones as dependencies. In case of monolith build, the function does nothing.
+# Example: networkit_module_link_modules(io graph) # io depends on graph
+function(networkit_module_link_modules modname)
+    if (NOT NETWORKIT_MONOLITH)
+        foreach(dep IN LISTS ARGN)
+            target_link_libraries(networkit_${modname} PUBLIC networkit_${dep})
+        endforeach()
+    endif()
+endfunction()
+
+################################################################################
+# TESTING and BENCHMARKING
+if (NETWORKIT_BUILD_TESTS)
+    enable_testing()
+
+    if(EXISTS "${PROJECT_SOURCE_DIR}/extlibs/googletest/CMakeLists.txt")
+        option(BUILD_GTEST "Builds the googletest subproject" ON)
+        option(BUILD_GMOCK "Builds the googlemock subproject" OFF)
+        add_subdirectory(extlibs/googletest)
+    else()
+        message(FATAL_ERROR
+            "Missing GoogleTest and GoogleMock in extlibs/googletest. "
+            "Please run `git submodule update --init` to fetch the submodule.")
+    endif()
+
+    if (NETWORKIT_MONOLITH)
+        add_executable(networkit_tests networkit/cpp/Unittests-X.cpp)
+        target_link_libraries(networkit_tests
+            PRIVATE
+                gtest_main gtest
+                networkit
+        )
+        set_target_properties(networkit_tests PROPERTIES
+            CXX_STANDARD 11
+            COMPILE_FLAGS "${NETWORKIT_CXX_FLAGS}"
+            LINK_FLAGS "${OpenMP_CXX_FLAGS}")
+
+        add_test(networkit_tests networkit_tests)
+		add_custom_target(symlink_input_folder ALL COMMAND ${CMAKE_COMMAND} -E create_symlink "${PROJECT_SOURCE_DIR}/input" "input")
```

Comment:
Warning for windows users is missing, indeed. 

Indentation is off here too, right. @manpen defaults his editor to spaces while I set mine to Tab.  

Of course they could run the networkit_tests executable from the networkit root directory and can make it without a symlink, but this is an odd behaviour when working with cmake because you expect it to work without moving the executable. The real problem is that ALL graph paths in the tests are given like `input/graph`. The only "clean" way to do this, would be changing the paths of the tests to be depended on a variable which would be the path from the build directory to the source directory.

Reply:
True, it is odd behavior. But it only affects potential Windows users who cannot even run NetworKit without patches right now anyway.

Reply:
It not necessary to move the executable? The way I currently run tests is using

```bash
cd build
[...]
make 
cd ..
build/networkit_tests -t
```
I think to more CMake-ish way to do it, is using CTest which allows to specify a working directory:

```cmake
 add_test(
                    NAME networkit_tests
                    COMMAND networkit_test -t
                    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
 )

 add_test(
                    NAME networkit_tests_no_assertions
                    COMMAND networkit_test -r
                    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
 )
```
and then invoke `ctest -V` after compilation.